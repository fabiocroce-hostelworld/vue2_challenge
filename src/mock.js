import { clavaAPI } from "@/api/clavaAPI"

clavaAPI.api.get = jest.fn()

export const mockAxiosGet = clavaAPI.api.get
