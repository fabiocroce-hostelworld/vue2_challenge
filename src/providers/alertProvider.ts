import { provide, inject, reactive, computed, ComputedRef } from "@vue/composition-api"

export enum alertTypes {
  SUCCESS = "success",
  ERROR = "error",
  INFO = "info",
  WARNING = "warning"
}

interface Alert {
  type: alertTypes
  message: string
  timerId?: number | undefined
}

export interface AlertProvider {
  addErrorAlert: (message: string) => void
  addSuccessAlert: (message: string) => void
  clearAlerts: () => void
  clearAlert: (idx: number) => void
  alerts: ComputedRef<Alert[]>
}

type alertConfig = { timeout?: number }

// Store for alert messages
export const setupAlertProvider = (config: alertConfig = {}): NonNullable<AlertProvider> => {
  const alertsArray = reactive<{ alerts: Array<Alert> }>({ alerts: [] })

  const clearByTimerId = (removeTimerId: number) => {
    alertsArray.alerts = alertsArray.alerts.filter(({ timerId }) => timerId !== removeTimerId)
  }

  const addAlert = ({ message, type }: Alert) => {
    if (config.timeout) {
      const timerId = setTimeout(() => {
        clearByTimerId(timerId)
      }, config.timeout)

      alertsArray.alerts.push({ message, type, timerId })
      return
    }
    alertsArray.alerts.push({ message, type })
  }

  const addErrorAlert = (message: string) => addAlert({ message, type: alertTypes.ERROR })
  const addSuccessAlert = (message: string) => addAlert({ message, type: alertTypes.SUCCESS })

  const clearAlerts = () => void (alertsArray.alerts.length = 0)
  const clearAlert = (removeIdx: number) => {
    alertsArray.alerts = alertsArray.alerts.filter((_, idx) => removeIdx !== idx)
  }
  const alerts = computed(() => alertsArray.alerts)

  return {
    addErrorAlert,
    addSuccessAlert,
    clearAlerts,
    clearAlert,
    alerts
  }
}

export const alertRef = Symbol("alert")

export const provideAlert = /* istanbul ignore next */ (config: alertConfig = {}) =>
  provide(alertRef, setupAlertProvider(config))
export const useAlert = () => inject<AlertProvider>(alertRef)
