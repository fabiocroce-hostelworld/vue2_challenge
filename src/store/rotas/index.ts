import { getRotaServices } from "@/api/clavaAPI"
import { IRotas } from "@/api/clavaAPI/rota/RotasResponseDTO"
import { IRotasList } from "@/api/clavaAPI/rotasUsers/RotasResponseDTO"
import { addNamespacePrefix } from "@/utils/vuexHelpers"

const types = {
  getters: {
    GET_ROTAS: "GET_ROTAS"
  },
  mutations: {
    STORE_ROTAS: "STORE_ROTAS",
    STORE_ROTA: "STORE_ROTA"
  },
  actions: {
    FETCH_ROTA: "fetchRota",
    GENERATE_ROTA: "generateRota"
  }
}

export const rotasTypes = addNamespacePrefix(types, "rotas")

type RotasState = {
  rotas: IRotasList
}

export default {
  namespaced: true,
  state: (): RotasState => ({
    rotas: [] as IRotasList
  }),
  getters: {
    [types.getters.GET_ROTAS](state: RotasState) {
      return state.rotas
    }
  },
  actions: {
    [types.actions.FETCH_ROTA]: async ({ commit }: { commit: Function }, rotaId: number) => {
      const rota = await getRotaServices().fetchRota(rotaId)

      commit(types.mutations.STORE_ROTA, rota)
    },
    [types.actions.GENERATE_ROTA]: async ({ dispatch }: { dispatch: Function }, date: string) => {
      const rotaId = await getRotaServices().generateRota(date)

      await dispatch(types.actions.FETCH_ROTA, rotaId)
    }
  },
  mutations: {
    [types.mutations.STORE_ROTAS]: (state: RotasState, rotas: IRotasList) => {
      state.rotas = rotas
    },
    [types.mutations.STORE_ROTA]: (state: RotasState, rota: IRotas) => {
      state.rotas.push(rota)
    }
  }
}
