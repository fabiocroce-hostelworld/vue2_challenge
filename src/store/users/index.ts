import { IUsersList } from "@/api/clavaAPI/rotasUsers/RotasResponseDTO"
import { addNamespacePrefix } from "@/utils/vuexHelpers"

const types = {
  getters: {
    GET_USERS: "getUsers",
    GET_MAPPED_USERS: "getMappedUsers",
    GET_USER_NAME: "getUserName"
  },
  mutations: {
    STORE_USERS: "STORE_USERS"
  },
  actions: {
    FETCH_USERS: "fetchUsers"
  }
}

export const usersTypes = addNamespacePrefix(types, "users")

type UsersState = {
  users: IUsersList
}

export default {
  namespaced: true,
  state: (): UsersState => ({
    users: [] as IUsersList
  }),
  getters: {
    [types.getters.GET_USERS](state: UsersState) {
      return state.users
    },
    [types.getters.GET_MAPPED_USERS](state: UsersState) {
      return Object.assign({}, ...state.users.map(user => ({ [user.userId]: user.user })))
    },
    [types.getters.GET_USER_NAME](_: UsersState, getters: { [x: string]: IUsersList }) {
      return (userId: number) => getters[types.getters.GET_MAPPED_USERS][userId]
    }
  },
  mutations: {
    [types.mutations.STORE_USERS]: (state: UsersState, users: IUsersList) => {
      state.users = users
    }
  }
}
