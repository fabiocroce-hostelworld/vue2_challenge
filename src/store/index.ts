import { getRotasUsersServices } from "@/api/clavaAPI"
import Vue from "vue"
import Vuex, { Store } from "vuex"
import rotas, { rotasTypes } from "@/store/rotas"
import users, { usersTypes } from "@/store/users"

Vue.use(Vuex)

export const rootTypes = {
  actions: {
    FETCH_ROTAS_AND_USERS: "fetchRotasAndUsers"
  }
}

export { rotasTypes, usersTypes }

export const storeConfig = {
  modules: {
    rotas,
    users
  },
  actions: {
    [rootTypes.actions.FETCH_ROTAS_AND_USERS]: async ({ commit }: { commit: Function }) => {
      const { users, rotas } = await getRotasUsersServices().fetchUsersAndRotas()

      commit(rotasTypes.mutations.STORE_ROTAS, rotas)
      commit(usersTypes.mutations.STORE_USERS, users)
    }
  }
}

export default new Store(storeConfig)
