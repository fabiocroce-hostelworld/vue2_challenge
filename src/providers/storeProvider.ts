import { provide, inject, computed, ComputedRef } from "@vue/composition-api"
import store from "@/store"

export interface StoreProvider {
  store: typeof store
  reactiveGetter: <T>(name: string) => ComputedRef<T>
}

// Vuex adapter for composition API
export const setupStoreProvider = (localStore: typeof store): StoreProvider => ({
  store: localStore,
  reactiveGetter: <T>(getterName: string): ComputedRef<T> => {
    return computed(() => localStore.getters[getterName])
  }
})

export const storeRef = Symbol("vuex-store")
export const provideStore = /* istanbul ignore next */ () => provide(storeRef, setupStoreProvider(store))
export const useStore = () => inject<StoreProvider>(storeRef)
