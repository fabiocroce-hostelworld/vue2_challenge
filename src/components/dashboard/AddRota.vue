<template>
  <v-card>
    <v-card-title> Add Route </v-card-title>
    <v-card-text>
      <v-form ref="form">
        <v-menu
          v-model="menu"
          :close-on-content-click="false"
          transition="scale-transition"
          offset-y
          max-width="290px"
          min-width="290px"
        >
          <template v-slot:activator="{ on }">
            <v-text-field label="From Date" prepend-icon="mdi-calendar" readonly :value="date" v-on="on" />
          </template>
          <v-date-picker :min="today" v-model="date" no-title @input="menu = false" />
        </v-menu>
        <v-btn :loading="isFetching" class="generate-route mr-4 primary" @click="handleClick(date)">
          Generate Route
        </v-btn>
      </v-form>
    </v-card-text>
  </v-card>
</template>

<script lang="ts">
import { ref } from "@vue/composition-api"
import { StoreProvider, useStore } from "@/providers/storeProvider"
import { rotasTypes } from "@/store"
import { AlertProvider, useAlert } from "@/providers/alertProvider"
import { useStatus } from "@/hooks/useStatus"
import Vue from "vue"

export default Vue.extend({
  name: "AddRota",
  props: {
    today: {
      type: String,
      required: false,
      default: new Date().toISOString().substr(0, 10)
    }
  },
  setup({ today }) {
    const { store } = useStore() as StoreProvider
    const { addSuccessAlert, addErrorAlert } = useAlert() as AlertProvider
    const { startFetching, resetFetch, isFetching } = useStatus()

    const date = ref(today)
    const menu = ref(false)

    const handleClick = async (selectedDate: string) => {
      startFetching()
      store
        .dispatch(rotasTypes.actions.GENERATE_ROTA, selectedDate)
        .then(() => void addSuccessAlert("Rota added succesfully."))
        .catch(() => void addErrorAlert("Could not add the rota for that date."))
        .finally(() => void resetFetch())
    }

    return {
      date: date.value,
      menu: menu.value,
      handleClick,
      isFetching
    }
  }
})
</script>
