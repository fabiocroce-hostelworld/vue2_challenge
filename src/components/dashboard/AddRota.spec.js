import Vuex from "vuex"
import { mount, createLocalVue } from "@vue/test-utils"
import { storeConfig, rotasTypes } from "@/store"
import AddRota from "@/components/dashboard/AddRota"
import rotaUsersMock from "@/api/clavaAPI/rotasUsers/rotasResponse.mock"
import { setupStoreProvider, storeRef } from "@/providers/storeProvider"
import { setupAlertProvider, alertRef } from "@/providers/alertProvider"
import { mockAxiosGet } from "@/mock"
import flushPromises from "flush-promises"

const alertProvider = setupAlertProvider({})
alertProvider.addErrorAlert = jest.fn()
alertProvider.addSuccessAlert = jest.fn()

describe("AddRota", () => {
  let input
  let localVue
  let store

  beforeEach(() => {
    localVue = createLocalVue()
    store = new Vuex.Store(storeConfig)

    input = {
      localVue,
      propsData: {
        today: new Date("2020-12-20").toISOString().substr(0, 10)
      },
      provide: {
        [storeRef]: setupStoreProvider(store),
        [alertRef]: alertProvider
      }
    }
  })

  it("Snapshot test", () => {
    const wrapper = mount(AddRota, input)

    expect(wrapper.html()).toMatchSnapshot()
  })

  it("test successful route generation", async () => {
    const wrapper = mount(AddRota, input)

    mockAxiosGet
      .mockResolvedValueOnce({ data: { status: "OK", rotaid: 13 }, status: 200 })
      .mockResolvedValueOnce({ data: rotaUsersMock.rotas[0], status: 200 })

    wrapper.find(".generate-route").trigger("click")
    await flushPromises()

    expect(alertProvider.addSuccessAlert).toHaveBeenCalled()
    expect(store.getters[rotasTypes.getters.GET_ROTAS]).toEqual([rotaUsersMock.rotas[0]])
  })

  it("test failed route generation", async () => {
    const wrapper = mount(AddRota, input)

    mockAxiosGet.mockRejectedValueOnce()

    wrapper.find(".generate-route").trigger("click")
    await flushPromises()

    expect(alertProvider.addErrorAlert).toHaveBeenCalled()
    expect(store.getters[rotasTypes.getters.GET_ROTAS]).toEqual([])
  })

  it("test failed route generation when status is not ok", async () => {
    const wrapper = mount(AddRota, input)

    mockAxiosGet.mockResolvedValueOnce({ data: { status: "NOT_FOUND" }, status: 404 })

    wrapper.find(".generate-route").trigger("click")
    await flushPromises()

    expect(alertProvider.addErrorAlert).toHaveBeenCalled()
    expect(store.getters[rotasTypes.getters.GET_ROTAS]).toEqual([])
  })
})
