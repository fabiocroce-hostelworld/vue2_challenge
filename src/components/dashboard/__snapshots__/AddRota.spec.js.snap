// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddRota Snapshot test 1`] = `
<div class="v-card v-sheet theme--light">
  <div class="v-card__title">
    Add Route
  </div>
  <div class="v-card__text">
    <form novalidate="novalidate"
          class="v-form"
    >
      <div class="v-menu">
        <div class="v-input v-input--is-label-active v-input--is-dirty v-input--is-readonly theme--light v-text-field">
          <div class="v-input__prepend-outer">
            <div class="v-input__icon v-input__icon--prepend">
              <i aria-hidden="true"
                 class="v-icon notranslate mdi mdi-calendar theme--light"
              >
              </i>
            </div>
          </div>
          <div class="v-input__control">
            <div class="v-input__slot">
              <div class="v-text-field__slot">
                <label for="input-10"
                       class="v-label v-label--active theme--light"
                       style="right: 0px; position: absolute;"
                >
                  From Date
                </label>
                <input id="input-10"
                       readonly="readonly"
                       type="text"
                >
              </div>
            </div>
            <div class="v-text-field__details">
              <div class="v-messages theme--light">
                <transition-group-stub tag="div"
                                       name="message-transition"
                                       class="v-messages__wrapper"
                >
                </transition-group-stub>
              </div>
            </div>
          </div>
        </div>
      </div>
      <button type="button"
              class="generate-route mr-4 primary v-btn v-btn--contained theme--light v-size--default"
      >
        <span class="v-btn__content">
          Generate Route
        </span>
      </button>
    </form>
  </div>
</div>
`;
